lib/forking/forked_storage_trie.js

```diff
@@ -12,119 +11,58 @@ inherits(ForkedStorageBaseTrie, BaseTrie);
-      // that doesn't require such checks
-      if (blockNumberProvided) {
-        // this logic is heavily influenced by BlockchainDouble.prototype.getStorage
-        // but some adjustments were necessary due to the ForkedStorageTrieBase context
-        self.blockchain.getBlock(blockNumber, function(err, block) {
+    if (typeof touchedAt !== "undefined") {
+      MerklePatriciaTree.prototype.get.call(self, key, function(err, r) {
+        callback(err, r);
+      });
+    } else {
+      // If this is the main trie, get the whole account.
+      if (self.address == null) {
+        self.blockchain.fetchAccountFromFallback(key, self.forkBlockNumber, function(err, account) {
           if (err) {
             return callback(err);
           }
 
-          // Manipulate the state root in place to maintain checkpoints
-          const currentStateRoot = self.root;
-          self.root = block.header.stateRoot;
-
-          MerklePatriciaTree.prototype.get.call(self, utils.toBuffer(self.address), function(err, data) {
-            if (err != null) {
-              // Put the stateRoot back if there's an error
-              self.root = currentStateRoot;
-              return callback(err);
-            }
-
-            const account = new Account(data);
-
-            self.root = account.stateRoot;
-            MerklePatriciaTree.prototype.get.call(self, key, function(err, value) {
-              // Finally, put the stateRoot back for good
-              self.root = currentStateRoot;
-
-              if (err != null) {
-                return callback(err, value);
-              }
-
-              callback(null, value);
-            });
-          });
+          callback(null, account.serialize());
         });
       } else {
-        MerklePatriciaTree.prototype.get.call(self, key, function(err, r) {
-          callback(err, r);
-        });
-      }
-    } else {
-      self.keyIsDeleted(key, (err, deleted) => {
-        if (err) {
-          return callback(err);
-        }
-
-        if (deleted) {
-          // it was deleted. return nothing.
-          callback(null, Buffer.allocUnsafe(0));
-          return;
-        }
-
-        // If this is the main trie, get the whole account.
-        if (self.address == null) {
-          self.blockchain.fetchAccountFromFallback(key, blockNumber, function(err, account) {
-            if (err) {
-              return callback(err);
-            }
-
-            callback(null, account.serialize());
-          });
-        } else {
-          if (to.number(blockNumber) > to.number(self.forkBlockNumber)) {
-            blockNumber = self.forkBlockNumber;
-          }
-          self.web3.eth.getStorageAt(to.rpcDataHexString(self.address), to.rpcDataHexString(key), blockNumber, function(
-            err,
-            value
-          ) {
+        self.web3.eth.getStorageAt(
+          to.rpcDataHexString(self.address),
+          to.rpcDataHexString(key),
+          self.forkBlockNumber,
```

Comment:
So I ran some hacked together tests to test a hunch I had about this `blockNumber` change. Here is the test:

```javascript
  it.only("gets values at specified blocks on the original change", async function() {
    const { send, accounts: [from], abi, web3: originalWeb3, instance: {_address: contractAddress}, provider: originalProvider } = forkedContext;
    const originalContract = new originalWeb3.eth.Contract(abi, contractAddress);

    const txParams = {from};
    
    await originalContract.methods.test().send(txParams);
    const initialValue = await originalContract.methods.value().call();
    await send("evm_mine", null);
    const {result: preForkBlockNumber} = await send("eth_blockNumber");

    // Fork the "original" chain _now_
    const { web3 } = await intializeTestProvider({
      fork: originalProvider,
      logger,
      seed: "forked provider"
    });

    const instance = new web3.eth.Contract(abi, contractAddress);

    // get the value as it was before we forked
    instance.defaultBlock = preForkBlockNumber;
    // there is a bug possibly unrelated to this PR that prevents this from
    // returning the wrong value (it crashes with a `pop of undefined` instead).
    const finalValue = await instance.methods.value().call();
    assert.equals(finalValue, initialValue);
  });
```

There is a problem with this test though... another bug seems to be preventing the final `await instance.methods.value().call();` from returning a value, due to a `'pop' of undefined` error.

The immediate issue I was investigating was the use of `self.forkBlockNumber` instead of `blockNumber`, but perhaps the issue is further up in the stack and we shouldn't be visiting this function anyway.

All that said, this change _might_ cause a regression, _but_ it also might not currently (i.e., `develop`) affect user code due to the other bug getting in the way. So, shipping it with a `// TODO` and a new Issue might not be the worst thing in the world. :sweat_smile:

Reply:
my fix seems to coincidentally remove the `'pop' of undefined` error 🤷